dwa_local_planner/src/dwa_planner_ros.cpp

```diff
@@ -244,18 +245,43 @@ namespace dwa_local_planner {
     base_local_planner::publishPlan(path, g_plan_pub_);
   }
 
-  void DWAPlannerROS::publishScaledFootprint(const geometry_msgs::PoseStamped& pose, const base_local_planner::Trajectory &traj) const {
-    visualization_msgs::Marker marker;
-    marker.header.frame_id = pose.header.frame_id;
-    marker.header.stamp = ros::Time::now();
-    marker.lifetime = ros::Duration(2 * dp_->getSimPeriod()); // double the sim period to avoid flickering
-    marker.type = visualization_msgs::Marker::LINE_STRIP;
-    marker.pose = pose.pose;
-    marker.scale.x = 0.01;
-    marker.color.g = marker.color.a = 1.0; // green
-    marker.points = dp_->getScaledFootprint(traj);
-    marker.points.push_back(marker.points.front()); // close the polygon
-    scaled_fp_pub_.publish(marker);
+  void DWAPlannerROS::publishScaledFootprint(const base_local_planner::Trajectory& traj) const {
+    if (scaled_fp_pub_.getNumSubscribers() == 0) {
+      return;
+    }
+
+    visualization_msgs::MarkerArray markers;
+
+    for (size_t i = 0; i < traj.getPointsSize(); ++i) {
+      visualization_msgs::Marker marker;
+      marker.header.frame_id = costmap_ros_->getGlobalFrameID();
+      marker.header.stamp = ros::Time::now();
+      marker.ns = i == 0 ? "current_footprint" : "future_footprints";
+      marker.id = i;
+      marker.lifetime = ros::Duration(2 * dp_->getSimPeriod());  // double the sim period to avoid flickering
+      marker.type = visualization_msgs::Marker::LINE_STRIP;
+
+      double yaw = 0;
+      traj.getPoint(i, marker.pose.position.x, marker.pose.position.y, yaw);
+      tf2::Quaternion quat_tf;
+      quat_tf.setRPY(0, 0, yaw);
+      marker.pose.orientation = tf2::toMsg(quat_tf);
+
+      // gradient: green -> yellow -> red
+      // https://stackoverflow.com/a/7947812
+      const double ratio = i / static_cast<double>(traj.getPointsSize() - 1);
+      marker.color.a = 1.0;
+      marker.color.r = std::min(2 * ratio, 1.0);
+      marker.color.g = std::min(2 * (1.0 - ratio), 1.0);
+      marker.scale.x = 0.01;
+
+      marker.points = dp_->getScaledFootprint(traj, i);
+      marker.points.push_back(marker.points.front());  // close the polygon
+
+      markers.markers.push_back(marker);
+    }
+
+    scaled_fp_pub_.publish(markers);
```

Comment:
That's a good question. I did some quick test recording both the old topic and the new one at the same time.
Seems like the new topic takes around 30~50 times more space?

```text
size		msgs		topic
------------------------------------------------
293.3K		404		old_topic
15.2M		404		new_topic

85.7K		118		old_topic
3.0M		118		new_topic

289.7K		399		old_topic
11.7M		399		new_topic
```

Using some random bagfiles I had on my pc as a reference, seems like the old topic usually takes around 1\~2 Mb.

So with this change I would expect to increase bag file size by 50\~100mb?
Maybe we can stop recording the topic since it's not very useful when investigating bagfiles anyways.

Reply:
> maybe it would be easier to tune if rather than a discount factor, one could instead e.g. specify the reduction factor at the end?

@renan028 had the same suggestion, so since you're both suggesting the same thing I implemented that.
Now the parameter defines the discount factor for the last point in the trajectory and we automatically compute the discount factor for the intermediate points.